text/0061-minimal-streams.md

```diff
@@ -0,0 +1,266 @@
+
+Data is transferred via the stream according to the following rules:
+1. On each cycle where both `valid` and `ready` are asserted, a transfer is performed: the contents of `payload` is conveyed from the transmitter to the receiver.
+2. Once the transmitter asserts `valid`, it must not deassert it until a transfer is performed.
+3. As an exception to rule (2), when the transmitter is reset it must deassert `valid`, unless it ties `valid` to `Const(1)` in the interface object.
+4. Once the transmitter asserts `valid`, it must not change the contents of `payload` until a transfer is performed.
+5. The transmitter must not assert `valid` in response to the receiver asserting `ready` (combinatorially or not).
+6. Once the receiver asserts `ready` it may deassert it at any time.
+7. The receiver may assert `ready` in response to the transmitter asserting `valid` (including combinatorially).
+
+These rules ensure that transmitters and receivers can be developed independently yet safely used together, without unduly restricting the application-specific conditions that determine assertion of `valid` and `ready`.
+
+Some transmitters and receivers may be designed in a way that requires them to perform a transfer on every cycle when they are not reset. Such transmitters and receivers may (but are not required to) tie `valid` to `Const(1)`. Similarly, some transmitters and receivers may be designed without support for backpressure. Such transmitters and receivers may (but are not required to) tie `ready` to `Const(1)`.
+
+If these control signals are tied to `Const(1)`, then `amaranth.lib.wiring.connect` ensures that only compatible streams are connected. For example, if the transmitter does not support backpressure (`ready` tied to `Const(1)`), it can only be connected to receivers that do not support backpressure. However, receivers that do not support backpressure can be connected to any transmitter, since the receiver would appear as an always-ready one to a transmitter that does support backpressure. Similar logic applies to `valid` when it is tied to `Const(1)`.
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new `amaranth.lib.stream` module is added, defining two classes: `Signature` and `Interface`:
+- `amaranth.lib.stream.Signature(payload_shape, *, always_valid=False, always_ready=False)`
+    - with members:
+        - `payload: Out(payload_shape)`
+        - `valid: Out(1)`
+        - `ready: In(1)`
+    - with properties:
+        - `always_valid: bool`
+        - `always_ready: bool`
+    - with methods:
+        - `__eq__(other)`: requires `isinstance(other, Signature)` and all three of `payload_shape`, `always_valid`, `always_ready` to match
```

Comment:
please clarify that it's `lib.stream.Signature` in particular

Reply:
done